### PR TITLE
Check if last directory name is same for localizing path

### LINKED ISSUE #24761: Load function breaks on filesystem paths that contain the full res:// path
**Godot version:**
latest Godot 3.1 alpha 5

**OS/device including version:**
Window 10 64 bit
GeForce GTX 1050 Ti

**Issue description:**
I am loading external tres mesh files and I have noticed when using unix pathing with / it does not properly load the resource because the full path is mangled.

Here is the test run using `"D:\\DEV\\godot\\DBdata\\mods/db/models/wall/block.tres"`
image

However `"D:/DEV/godot/DBdata/mods/db/models/wall/block.tres"` the object is null
image
And in fact gives an error
image

The error is 
```
E 0:00:00:0530   Failed loading resource: res:/data/mods/db/models/wall/block.tres
  <C Source>     core/io/resource_loader.cpp:279 @ _load()
  <Stack Trace>  preload.gd:76 @ _ready()
```
If you look closer you notice the path that is looking for somehow has become `res:/data/mods/db/models/wall/block.tres`

The local path of the project is `D:/DEV/godot/DB`

I believe what is going on is when using Unix type pathing the local path is cut out of the load path and the final load path becomes `"res://" + original path - local path` (pseudocode)

**Steps to reproduce:**

**Minimal reproduction project:**

### PR DESCRIPTION
**Pull request by volzhs**
Fix #24761

### CODE REVIEW

File: core/project_settings.cpp [hunk 1]
```diff
@@ -79,6 +79,12 @@ String ProjectSettings::localize_path(const String &p_path) const {
 			return p_path;
 		};
 
+		Vector<String> res_arr = resource_path.split("/");
+		Vector<String> cwd_arr = cwd.split("/");
+		if (res_arr[res_arr.size() - 1] != cwd_arr[res_arr.size() - 1]) {
```
Comment (akien-mga) [inline core/project_settings.cpp]: I don't really understand what this check does. A comment would be good to explain why this condition means that we should return the unchanged `p_path`.

Comment (volzhs) [reply core/project_settings.cpp]: @akien-mga added comment

**Comment by akien-mga** [comment]
I did a quick test in GDScript and I think this code can fail in some situations, especially if `resource_path` ends with `/`, as `split("/")` would then have an empty string as last item:
```gdscript
	var res_path = "/mnt/project/my_a"
	var res_path2 = "/mnt/project/my_a/"
	
	var test_path = "/mnt/project/my_a_dev"
	var test_path2 = "/mnt/project/my_a_dev/"
	var test_path3 = "/mnt/project/my_a/dev"
	var test_path4 = "/mnt/project/my_a/dev/"
	
	var res_arr = res_path.split("/")
	var res_arr2 = res_path2.split("/")
	var test_arr = test_path.split("/")
	var test_arr2 = test_path2.split("/")
	var test_arr3 = test_path3.split("/")
	var test_arr4 = test_path4.split("/")
	"""
	[, mnt, project, my_a]
	[, mnt, project, my_a, ]
	[, mnt, project, my_a_dev]
	[, mnt, project, my_a_dev, ]
	[, mnt, project, my_a, dev]
	[, mnt, project, my_a, dev, ]
	"""
	
	print(res_arr[res_arr.size() - 1] != test_arr[res_arr.size() - 1])  # True
	#print(res_arr2[res_arr2.size() - 1] != test_arr[res_arr2.size() - 1])  # invalid get index '4'
	print(res_arr[res_arr.size() - 1] != test_arr2[res_arr.size() - 1])  # True
	print(res_arr2[res_arr2.size() - 1] != test_arr2[res_arr2.size() - 1])  # False
	print(res_arr[res_arr.size() - 1] != test_arr3[res_arr.size() - 1])  # False
	print(res_arr2[res_arr2.size() - 1] != test_arr3[res_arr2.size() - 1])  # True
	print(res_arr[res_arr.size() - 1] != test_arr4[res_arr.size() - 1])  # False
	print(res_arr2[res_arr2.size() - 1] != test_arr4[res_arr2.size() - 1])  # True
```

It works better with `split("/", false)` to ignore empty strings. But I wonder if it's the most efficient way to check whether `p_path` contains `resource_path`. An alternative could be to check if `p_path` contains `resource_path + "/"` maybe?

**Comment by akien-mga** [comment]
This seems to work:
```
if (!cwd.begins_with(resource_path.plus_file("")))
	return p_path;
}
```
`plus_file("")` ensures that the path ends with a `/`.

Tested in GDScript and it seems to work fine:
```gdscript
	var res_path = "/mnt/project/my_a"
	var res_path2 = "/mnt/project/my_a/"
	
	var test_path = "/mnt/project/my_a_dev"
	var test_path2 = "/mnt/project/my_a_dev/"
	var test_path3 = "/mnt/project/my_a/dev"
	var test_path4 = "/mnt/project/my_a/dev/"
	
	print(res_path.plus_file(""))  # "/mnt/project/my_a"
	print(res_path2.plus_file(""))  # "/mnt/project/my_a"
	
	print(!test_path.begins_with(res_path.plus_file("")))  # True
	print(!test_path.begins_with(res_path2.plus_file("")))  # True
	print(!test_path2.begins_with(res_path.plus_file("")))  # True
	print(!test_path2.begins_with(res_path2.plus_file("")))  # True
	print(!test_path3.begins_with(res_path.plus_file("")))  # False
	print(!test_path3.begins_with(res_path2.plus_file("")))  # False
	print(!test_path4.begins_with(res_path.plus_file("")))  # False
	print(!test_path4.begins_with(res_path2.plus_file("")))  # False
```

**Comment by akien-mga** [comment]
This can even be a bit better, I'll do another proposal PR and some tests.

**Comment by volzhs** [comment]
@akien-mga force-pushed with your suggestion. it's up to you. :smile:

**Comment by akien-mga** [comment]
Ah, here's the alternative commit I was about to propose. We can actually check both conditions at once and replace with `res://` instead of the awkward `res:/`:

```diff
commit 9ad9d1f3b3a5e6d1440a0f8d38e1b3246e00412d
Author: Rémi Verschelde <rverschelde@gmail.com>
Date:   Thu May 30 16:03:12 2019 +0200

    Don't localize paths that contain but are not in the resource path
    
    This issue could be triggered if you try to access a path which contains
    the resource path string in its absolute path, while pointing to a directory
    which is *not* in the resource path.
    
    It's clearer with an example: with `/my/project` as resource path, the
    previous logic would also localize `/my/project_data` to `res://data`, which
    is incorrect and would lead to a cryptic error.
    
    Fixes #24761.
    
    Co-authored-by: volzhs <volzhs@gmail.com>

diff --git a/core/project_settings.cpp b/core/project_settings.cpp
index 2fd22d478..4c37142ff 100644
--- a/core/project_settings.cpp
+++ b/core/project_settings.cpp
@@ -75,11 +75,19 @@ String ProjectSettings::localize_path(const String &p_path) const {
 
 		memdelete(dir);
 
-		if (!cwd.begins_with(resource_path)) {
+		// Ensure that we end with a '/'.
+		// This is important to ensure that we do not wrongly localize the resource path
+		// in an absolute path that just happens to contain this string but points to a
+		// different folder (e.g. "/my/project" as resource_path would be contained in
+		// "/my/project_data", even though the latter is not part of res://.
+		// `plus_file("")` is an easy way to ensure we have a trailing '/'.
+		const String res_path = resource_path.plus_file("");
+
+		if (!cwd.begins_with(res_path)) {
 			return p_path;
 		};
 
-		return cwd.replace_first(resource_path, "res:/");
+		return cwd.replace_first(res_path, "res://");
 	} else {
 
 		memdelete(dir);
```

**Comment by volzhs** [comment]
@akien-mga I would close mine then. go ahead. :+1:

**Comment by akien-mga** [comment]
Ok, I opened #29315 which supersedes this PR. Thanks for the fix :)

**Pull request closed without merging**